godot-core/src/builtin/rid.rs

```diff
@@ -0,0 +1,140 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::num::NonZeroU64;
+
+use godot_ffi as sys;
+use sys::{ffi_methods, GodotFfi};
+
+/// A RID ("resource ID") is an opaque handle that refers to a Godot `Resource`.
+///
+/// RIDs do not grant access to the resource itself. Instead, they can be used in lower-level resource APIs
+/// such as the [servers]. See also [Godot API docs for `RID`][docs].
+///
+/// RIDs should be largely safe to work with. Certain calls to servers may fail, however doing so will
+/// trigger an error from Godot, and will not cause any UB.
+///
+/// # Layout:
+///
+/// `Rid` has the same layout as a [`u64`], where [`Rid::Invalid`] is 0, and [`Rid::Valid(i)`] is `i`.
```

Comment:
I don't think we should expose this implementation detail, it only limits us in the future (e.g. if we ever add a debug-only field for extra validation or so). We already have `new` and `to_u64` functions that are reasonably cheap.

Reply:
Also, as far as I know, without `#[repr(C)]` and `#[repr(transparent)]`, no layout is guaranteed, even if with the enum optimization you mentioned. It's very unlikely to change, but a compiler is theoretically free to do so.

Reply:
> Also, as far as I know, without `#[repr(C)]` and `#[repr(transparent)]`, no layout is guaranteed, even if with the enum optimization you mentioned. It's very unlikely to change, but a compiler is theoretically free to do so.

The rustnomicon explicitly says the nullable pointer optimization is guaranteed:
> This is called an "optimization", but unlike other optimizations it is guaranteed to apply to eligible types.

Reply:
I also just ran the itests with `-Zrandomize-layout` and it works. If the layout wasn't guaranteed here, then `rid_equiv` should fail under that. Since it checks that `Rid::Invalid` maps to 0, and `Rid::Valid(i)` maps to `i` (for one specific `i` at least)

Reply:
`-Zrandomize-layout` passing is a necessary but not sufficient criterion for correct layout -- it only reorders fields and doesn't change fields etc. But to my understanding, it's conforming for a Rust compiler to e.g. add padding bytes at the end, even if the enum optimization is available. But we're really talking about theory here, I also don't see why this would happen.

Anyway, it doesn't matter much: the point is that a guaranteed layout limits potential implementation changes in the future, so we shouldn't document it. There's the existing API for such conversions.